dll/win32/shell32/CDefViewUtil.cpp

```diff
@@ -0,0 +1,132 @@
+/*
+ * PROJECT:     shell32
+ * LICENSE:     LGPL-2.1+ (https://spdx.org/licenses/LGPL-2.1+)
+ * PURPOSE:     DefView utility functions
+ * COPYRIGHT:   Copyright 2024 Whindmar Saksit <whindsaks@proton.me>
+ */
+
+#include "precomp.h"
+
+WINE_DEFAULT_DEBUG_CHANNEL(shell);
+
+class CObjectWithSiteBase :
+    public IObjectWithSite
+{
+    public:
+        IUnknown* m_pUnkSite;
+
+        CObjectWithSiteBase() : m_pUnkSite(NULL) {}
+        virtual ~CObjectWithSiteBase() { SetSite(NULL); }
+
+        // IObjectWithSite
+        STDMETHODIMP SetSite(IUnknown *pUnkSite) override
+        {
+            IUnknown_Set(&m_pUnkSite, pUnkSite);
+            return S_OK;
+        }
+        STDMETHODIMP GetSite(REFIID riid, void **ppvSite) override
+        {
+            *ppvSite = NULL;
+            return m_pUnkSite ? m_pUnkSite->QueryInterface(riid, ppvSite) : E_FAIL;
+        }
+};
+
+// This class adapts the legacy function callback to work as an IShellFolderViewCB
+class CShellFolderViewCBWrapper :
+    public CComObjectRootEx<CComMultiThreadModelNoCS>,
+    public IShellFolderViewCB,
+    public CObjectWithSiteBase
+{
+    protected:
+        HWND                    m_hWndMain;
+        CComPtr<IShellFolder>   m_psf;
+        CComPtr<IShellView>     m_psvOuter;
+        LPFNVIEWCALLBACK        m_Callback;
+        FOLDERVIEWMODE          m_FVM;
+        LONG                    m_Events;
+    public:
+        CShellFolderViewCBWrapper() : m_hWndMain(NULL) {}
+
+        HRESULT WINAPI Initialize(LPCSFV psvcbi)
+        {
+            m_psf = psvcbi->pshf;
+            m_psvOuter = psvcbi->psvOuter;
+            m_Callback = psvcbi->pfnCallback;
+            m_FVM = psvcbi->fvm;
+            m_Events = psvcbi->lEvents;
+            return S_OK;
+        }
+
+        // IShellFolderViewCB
+        STDMETHODIMP MessageSFVCB(UINT uMsg, WPARAM wParam, LPARAM lParam) override
+        {
+            switch (uMsg)
+            {
+                case SFVM_HWNDMAIN:
+                    m_hWndMain = (HWND)lParam;
+                    break;
+
+                case SFVM_DEFVIEWMODE:
+                    if (m_FVM)
+                        *(FOLDERVIEWMODE*)lParam = m_FVM;
+                    break;
+            }
+
+            HRESULT hr = m_Callback(m_psvOuter, m_psf, m_hWndMain, uMsg, wParam, lParam);
+            if (SUCCEEDED(hr))
+                return hr;
+
+            switch (uMsg)
+            {
+                case SFVM_GETNOTIFY:
+                    *(LPITEMIDLIST*)wParam = NULL; // TODO: Where is this pidl supposed to come from?
```

Comment:
(Note for reviewers: https://learn.microsoft.com/en-us/windows/win32/shell/sfvm-getnotify#parameters )
> A pointer to an absolute IDList of an item for which the view should register to be notified of changes. Typically, this is the same as the IDList of the location being viewed, but it can be another location.

Is it possible this is somehow obtained from the CDefView just after the SHCreateShellFolderView() call done in the Ex version (and before returning)?

Reply:
This is after the callback has already failed, if the caller wants this to be dynamic they would have already done so in their callback.

The `SHCreateShellFolderViewEx` caller can pass a PIDL in `LPCSFV` but MSDN says this PIDL is ignored. If that was true in 95, I don't know and I don't really care to find out unless an actual bug appears.

Reply:
I believe the current code in our CDefView is incorrect. MSDN says for SFVM_GETNOTIFY 
 > If this callback message does not return a nonzero value for either the IDList **or** the events mask then the view will not register for change notifications.

but in this case ROS registers SHCNE_ALLEVENTS for m_pidlParent. It seems to me like the PIDL being NULL means all folders when the callback only sets the events variable.

Reply:
hmhm... Maybe @yagoulas has an idea about this :thinking: